File: frontend/src/stores/auth.store.ts
import { create } from 'zustand'
import { persist } from 'zustand/middleware'
import { User, LoginRequest, RegisterRequest } from '../types/api.types'
import { apiService } from '../services/api.service'

interface AuthState {
  user: User | null
  isAuthenticated: boolean
  isLoading: boolean
  error: string | null
  
  // Actions
  login: (credentials: LoginRequest) => Promise<void>
  register: (data: RegisterRequest) => Promise<void>
  logout: () => void
  clearError: () => void
  updateUser: (userData: Partial<User>) => void
  refreshUser: () => Promise<void>
}

export const useAuthStore = create<AuthState>()(
  persist(
    (set, get) => ({
      user: null,
      isAuthenticated: false,
      isLoading: false,
      error: null,

      login: async (credentials: LoginRequest) => {
        set({ isLoading: true, error: null })
        
        try {
          const response = await apiService.login(credentials)
          set({
            user: response.user,
            isAuthenticated: true,
            isLoading: false,
            error: null,
          })
        } catch (error) {
          set({
            isLoading: false,
            error: error instanceof Error ? error.message : 'Login failed',
            isAuthenticated: false,
            user: null,
          })
          throw error
        }
      },

      register: async (data: RegisterRequest) => {
        set({ isLoading: true, error: null })
        
        try {
          await apiService.register(data)
          // After successful registration, automatically log in
          await get().login({
            email: data.email,
            password: data.password,
          })
        } catch (error) {
          set({
            isLoading: false,
            error: error instanceof Error ? error.message : 'Registration failed',
          })
          throw error
        }
      },

      logout: () => {
        apiService.logout().catch(console.error)
        set({
          user: null,
          isAuthenticated: false,
          error: null,
        })
      },

      clearError: () => {
        set({ error: null })
      },

      updateUser: (userData: Partial<User>) => {
        const currentUser = get().user
        if (currentUser) {
          set({
            user: { ...currentUser, ...userData },
          })
        }
      },

      refreshUser: async () => {
        if (!get().isAuthenticated) return
        
        try {
          const user = await apiService.getCurrentUser()
          set({ user })
        } catch (error) {
          console.error('Failed to refresh user:', error)
          // If refresh fails, user might need to re-authenticate
          get().logout()
        }
      },
    }),
    {
      name: 'auth-storage',
      partialize: (state) => ({
        user: state.user,
        isAuthenticated: state.isAuthenticated,
      }),
    }
  )
)

File: spendly/ios/Spendly/ViewModels/AuthStore.swift
import Foundation
import SwiftUI
import Combine

@MainActor
class AuthStore: ObservableObject {
    @Published var user: User?
    @Published var isAuthenticated = false
    @Published var isLoading = false
    @Published var error: String?
    
    private let apiService = APIService.shared
    
    init() {
        checkAuthStatus()
    }
    
    func checkAuthStatus() {
        if KeychainManager.shared.getToken() != nil {
            Task {
                await refreshUser()
            }
        }
    }
    
    func login(email: String, password: String) async {
        isLoading = true
        error = nil
        
        do {
            let (_, user) = try await apiService.login(email: email, password: password)
            self.user = user
            self.isAuthenticated = true
        } catch {
            self.error = error.localizedDescription
        }
        
        isLoading = false
    }
    
    func register(firstName: String, lastName: String, email: String, password: String, currency: String) async {
        isLoading = true
        error = nil
        
        let request = RegisterRequest(
            email: email,
            password: password,
            firstName: firstName,
            lastName: lastName,
            defaultCurrency: currency
        )
        
        do {
            _ = try await apiService.register(request)
            // Auto-login after registration
            await login(email: email, password: password)
        } catch {
            self.error = error.localizedDescription
        }
        
        isLoading = false
    }
    
    func logout() async {
        isLoading = true
        
        do {
            try await apiService.logout()
        } catch {
            // Even if API logout fails, clear local state
        }
        
        user = nil
        isAuthenticated = false
        isLoading = false
    }
    
    func refreshUser() async {
        do {
            user = try await apiService.getCurrentUser()
            isAuthenticated = true
        } catch {
            // If refresh fails, user needs to login again
            isAuthenticated = false
        }
    }
    
    func updateProfile(_ updatedUser: User) async {
        isLoading = true
        error = nil
        
        do {
            user = try await apiService.updateProfile(updatedUser)
        } catch {
            self.error = error.localizedDescription
        }
        
        isLoading = false
    }
}